### PR TITLE
Event broadcasting system

### LINKED ISSUE #694: Provide a way for multiple systems to subscribe to message broadcasts
Things like `DocumentIsDirty` shouldn't have to be carefully routed to its many concerned parties. Those parties should subscribe to a `DocumentIsDirty` event.

### PR DESCRIPTION
**Pull request by mfish33**
Closes #694

### CODE REVIEW

File: editor/src/viewport_tools/tool.rs
Comment (Keavon) [inline editor/src/viewport_tools/tool.rs:20]: Make these options.

**Comment by TrueDoctor** [comment]
Im a bit confused about the naming of the pr, this system in in no way pull based right?
I mean we never do
```rust
loop{ 
   has varible changed();
}
```

File: editor/src/communication/broadcast_message.rs
Comment (Keavon) [reply editor/src/communication/broadcast_message.rs:11]: Max and I discussed this, I advocated for avoidance of the word "immediate" because it implies that it happens, well, immediately. But that isn't true: it puts it on the front of the queue and then perhaps others in this, or a function elsewhere in the callstack, may enqueue more stuff at the front before it finally returns all the way back to the dispatcher and then executes them in order of the queue (possibly triggering others to insert more messages first in the queue also). So it really isn't immediate.

Using that word also reinforces the misconception, which was already a point of confusion for past contributor, that a message added to the queue happens right away rather than waiting until the entire function (and any ancestor functions in the call stack) return, because it works by mutating the queue and eventually returning back to the dispatcher rather than working in the more traditional manner of an immediate function call.

"Delayed" is also confusing since it seems to imply some time delay, rather than being put at the back of the message queue. I'm certainly open to more naming suggestions but I think this is a pretty reasonable one that I'm happy with.

File: editor/src/communication/dispatcher.rs
Comment (Keavon) [reply editor/src/communication/dispatcher.rs:49]: How? It's a discriminant of a variant of a sub-message. This was the only way Max and I could figure out how to make it work. I do kind of like the cleanness of the separation though.

File: editor/src/communication/broadcast_message.rs
Comment (Keavon) [reply editor/src/communication/broadcast_message.rs:17]: Agreed!

File: editor/src/viewport_tools/tool.rs
Comment (mfish33) [reply editor/src/viewport_tools/tool.rs:44]: I honestly feel like it is less of a special case. When writing tools I think it feels far less "magical" than it did before. Instead of having to map these messages in a different file, almost everything regarding a tool is in one location. In addition, it removes some of the complex piping previously required in `tool_message_handler.rs`

**Comment by mfish33** [comment]
> Im a bit confused about the naming of the pr, this system in in no way pull based right? I mean we never do
> 
> ```rust
> loop{ 
>    has varible changed();
> }
> ```

I mean yes we never loop for a change but in a sense, it is still pull-based. Message handlers choose what to listen to at any given time.

File: editor/src/viewport_tools/tool.rs
Comment (mfish33) [reply editor/src/viewport_tools/tool.rs:103]: I fixed this by creating a closure for each case, subscribing, and unsubscribing.

**Review by Keavon — APPROVED** [review]
Good to merge after @TrueDoctor approval.

File: editor/src/communication/broadcast_message.rs
Comment (TrueDoctor) [reply editor/src/communication/broadcast_message.rs:11]: The second point still stands though, I think we only need one of the messages anyway, making the naming discussion obsolete

File: editor/src/communication/dispatcher.rs
Comment (TrueDoctor) [reply editor/src/communication/dispatcher.rs:49]: The same way that you would deduplicate any other message
```rust
	MessageDiscriminant::Broadcast(BroadcastMessageDiscriminant::UpdateDocumentLayer),

```

File: editor/src/viewport_tools/tool.rs
Comment (TrueDoctor) [reply editor/src/viewport_tools/tool.rs]: You can mark this as resolved when you push the fix

File: editor/src/communication/broadcast_message.rs
Comment (Keavon) [reply editor/src/communication/broadcast_message.rs:11]: Ah, you're right, thanks for pointing that out. I understand that you mean now.

File: editor/src/viewport_tools/tool.rs
Comment (TrueDoctor) [reply editor/src/viewport_tools/tool.rs]: Sure was just a gut feeling

File: editor/src/communication/broadcast_message.rs
Comment (mfish33) [reply editor/src/communication/broadcast_message.rs:23]: Yes, I think this is true but it should be out of the scope of this PR. currently, the properties panel only updates on selection changed not document is dirty.

File: editor/src/document/properties_panel_message_handler.rs
Comment (TrueDoctor) [inline editor/src/document/properties_panel_message_handler.rs:162]: I think the into() should take care of the boxing so you don't need to write that out explicitly.
But I'm not sure if that would require type hints so feel free to ignore

**Review by TrueDoctor — APPROVED** [review]
Nice work looks good to me

**Pull request merged**
